sw/index.js

```diff
@@ -1,2 +1,92 @@
+
+precacheAndRoute(manifest)
+
+// Using network-only as the default strategy ensures that we fallback
+// to the browser as if the service worker wouldn't exist.
+// The browser may use own caching (HTTP cache).
+// Also, the offline fallback only works if request matched a route
+setDefaultHandler(new NetworkOnly())
+
+// This won't work in dev because pages are never cached.
+// See https://github.com/vercel/next.js/blob/337fb6a9aadb61c916f0121c899e463819cd3f33/server/render.js#L181-L185
+offlineFallback({ pageFallback: '/offline' })
+
+self.addEventListener('push', async function (event) {
+  const payload = event.data?.json()
+  if (!payload) return
+  const { tag } = payload.options
+  event.waitUntil((async () => {
+    if (!['REPLY', 'MENTION'].includes(tag)) {
+      return self.registration.showNotification(payload.title, payload.options)
+    }
+
+    const notifications = await self.registration.getNotifications({ tag })
+    // since we used a tag filter, there should only be zero or one notification
+    if (notifications.length > 1) {
+      console.error(`more than one notification with tag ${tag} found`)
+      return null
+    }
+    if (notifications.length === 0) {
+      return self.registration.showNotification(payload.title, payload.options)
+    }
+    const currentNotification = notifications[0]
+    const amount = currentNotification.data?.amount ? currentNotification.data.amount + 1 : 2
+    let title = ''
+    if (tag === 'REPLY') {
+      title = `You have ${amount} new replies`
+    } else if (tag === 'MENTION') {
+      title = `You were mentioned ${amount} times`
+    }
+    currentNotification.close()
+    const { icon } = currentNotification
+    return self.registration.showNotification(title, { icon, tag, data: { url: '/notifications', amount } })
+  })())
+})
+
+self.addEventListener('notificationclick', (event) => {
+  const url = event.notification.data?.url
+  if (url) {
+    event.waitUntil(self.clients.openWindow(url))
+  }
+})
+
+self.addEventListener('pushsubscriptionchange', (event) => {
+  // https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerGlobalScope/pushsubscriptionchange_event
+  const query = `
+  mutation savePushSubscription($endpoint: String!, $p256dh: String!, $auth: String!, $oldEndpoint: String!) {
+    savePushSubscription(endpoint: $endpoint, p256dh: $p256dh, auth: $auth, oldEndpoint: $oldEndpoint) {
+      id
+    }
+  }`
+  const subscription = self.registration.pushManager
+    .subscribe(event.oldSubscription.options)
+    .then((subscription) => {
+      // convert keys from ArrayBuffer to string
+      subscription = JSON.parse(JSON.stringify(subscription))
+      const variables = {
+        endpoint: subscription.endpoint,
+        p256dh: subscription.keys.p256dh,
+        auth: subscription.keys.auth,
+        oldEndpoint: event.oldSubscription.endpoint
+      }
+      const body = JSON.stringify({ query, variables })
+      return fetch('/api/graphql', {
```

Comment:
Regarding this:

> Although examples demonstrating how to share subscription related information with the application server tend to use [fetch()](https://developer.mozilla.org/en-US/docs/Web/API/fetch), this is not necessarily the best choice for real-world use, since it will not work if the app is offline, for example.
>
> Consider using another method to synchronize subscription information between your service worker and the app server, or make sure your code using fetch() is robust enough to handle cases where attempts to exchange data fail.

-- https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerGlobalScope/pushsubscriptionchange_event#usage_notes

I have read somewhere that the [Background Sync API](https://developer.chrome.com/blog/background-sync/) is recommended to use. However, I would keep it as it is now since the worst case that happens is that a user no longer receives push notifications at some point but they can easily resubscribe by checking the button in /notifications again. Also, I would like to not make the code any more complicated in the first iteration. We can make this event listener more stable in next iterations. Then I should know more about how to use Background Sync API and what the gotchas are.